src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComClassGenerator.cs

```diff
@@ -0,0 +1,225 @@
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.Interop
+{
+    [Generator]
+    public class ComClassGenerator : IIncrementalGenerator
+    {
+        private sealed record ComClassInfo(string ClassName, ContainingSyntaxContext ContainingSyntaxContext, ContainingSyntax ClassSyntax, SequenceEqualImmutableArray<string> ImplementedInterfacesNames);
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            // Get all types with the [GeneratedComClassAttribute] attribute.
+            var attributedClasses = context.SyntaxProvider
+                .ForAttributeWithMetadataName(
+                    TypeNames.GeneratedComClassAttribute,
+                    static (node, ct) => node is ClassDeclarationSyntax,
+                    static (context, ct) =>
+                    {
+                        var type = (INamedTypeSymbol)context.TargetSymbol;
+                        var syntax = (ClassDeclarationSyntax)context.TargetNode;
+                        ImmutableArray<string>.Builder names = ImmutableArray.CreateBuilder<string>();
+                        foreach (INamedTypeSymbol iface in type.AllInterfaces)
+                        {
+                            if (iface.GetAttributes().Any(attr => attr.AttributeClass?.ToDisplayString() == TypeNames.GeneratedComInterfaceAttribute))
+                            {
+                                names.Add(iface.ToDisplayString());
+                            }
+                        }
+                        return new ComClassInfo(
+                            type.ToDisplayString(),
+                            new ContainingSyntaxContext(syntax),
+                            new ContainingSyntax(syntax.Modifiers, syntax.Kind(), syntax.Identifier, syntax.TypeParameterList),
+                            new(names.ToImmutable()));
+                    });
+
+            var classInfoType = attributedClasses.Select(static (info, ct) => GenerateClassInfoType(info)).SelectNormalized();
+
+            var attribute = attributedClasses.Select(static (info, ct) => GenerateClassInfoAttributeOnUserType(info)).SelectNormalized();
+
+            context.RegisterSourceOutput(attributedClasses.Zip(classInfoType).Zip(attribute), static (context, classInfo) =>
+            {
+                var ((comClassInfo, classInfoType), attribute) = classInfo;
+                StringWriter writer = new();
+                writer.WriteLine(classInfoType.ToFullString());
+                writer.WriteLine();
+                writer.WriteLine(attribute);
+                context.AddSource(comClassInfo.ClassName, writer.ToString());
+            });
+        }
+
+        private const string ClassInfoTypeName = "ComClassInformation";
```

Comment:
Should this be in our "types" constants?

Reply:
So far, our "types" constants are for types in the BCL that we use or for types that we are introducing into the BCL. We currently don't have any type names that we generate that are not part of any public surface area in that type.